activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -81,19 +81,25 @@ def saved_change_to_target?
 
       private
         def replace(record)
+          reset_inverse_instance(target)
+
           if record
             raise_on_type_mismatch!(record)
             set_inverse_instance(record)
             @updated = true
-          elsif target
```

Comment:
Calling `reset_inverse_instance(target)` above takes care also of this case in a more generalised way.